src/win/fs.c

```diff
@@ -1611,26 +1613,356 @@ static void fs__access(uv_fs_t* req) {
+   */
+  DWORD numGroups = 0;
+  HANDLE hToken = NULL;
+  DWORD tokenAccess = TOKEN_IMPERSONATE |
+                      TOKEN_QUERY |
+                      TOKEN_DUPLICATE |
+                      STANDARD_RIGHTS_READ;
+  if (!OpenProcessToken(GetCurrentProcess(), tokenAccess, &hToken )) {
+    SET_REQ_WIN32_ERROR(req, GetLastError());
+    goto chmod_cleanup;
+  }
+  HANDLE hImpersonatedToken = NULL;
+  if (!DuplicateToken(hToken, SecurityImpersonation, &hImpersonatedToken)) {
+    SET_REQ_WIN32_ERROR(req, GetLastError());
+    goto chmod_cleanup;
+  }
+
+  /* Extract EAs from old DACL */
+  ULONG numOldEAs = 0;
+  ULONG numOtherGroups = 0;
+  if (ERROR_SUCCESS != GetExplicitEntriesFromAclW(pOldDACL, &numOldEAs, &poldEAs)) {
+    SET_REQ_WIN32_ERROR(req, GetLastError());
+    goto chmod_cleanup;
+  }
+
+  /* Iterate over all old ACEs, looking for groups that we belong to */
+  for (ULONG ea_idx=0; ea_idx<numOldEAs; ++ea_idx) {
+    PSID pEASid = (PSID)poldEAs[ea_idx].Trustee.ptstrName;
+    /* Skip this EA if it isn't an SID, or it is "Everyone" or our actual group */
+    if (poldEAs[ea_idx].Trustee.TrusteeForm != TRUSTEE_IS_SID ||
+        EqualSid(pEASid, psidEveryone) ||
+        EqualSid(pEASid, psidGroup)) {
+      continue;
+    }
+
+    /* Check to see if our user is a member of this group */
+    BOOL isMember = FALSE;
+    if (!CheckTokenMembership(hImpersonatedToken, pEASid, &isMember)) {
+      SET_REQ_WIN32_ERROR(req, GetLastError());
+      goto chmod_cleanup;
+    }
+
+    /* If we're a member, then count it */
+    if (isMember) {
+      numOtherGroups++;
+    }
+  }
+
+  /*
+   * Create an ACE for each triplet (user, group, other).
+   * Note that there is a priority mismatch between POSIX permissions and Windows ACLS:
+   * POSIX permissions are specificity-primary, in that a more specific permission takes
+   * precedence over a less specific one.  E.g. 0600 (allow user but deny group) will still
+   * end up allowing, whereas 0060 (deny user but allow group) will end up denying, because
+   * the "user" triplet is more specific than the group triplet.
+   * 
+   * In Windows ACLs, it is deny-primary; in that a deny anywhere in the ACL will result
+   * in a deny; therefore we build our masks by OR-ing in the more specific permissions with
+   * the less-specific permissions.
+   */
+  ea = (PEXPLICIT_ACCESS_W) malloc(sizeof(EXPLICIT_ACCESS_W)*(6 + 2*numOtherGroups));
+  DWORD u_mode = (req->fs.info.mode & S_IRWXU) >> 6;
+  DWORD g_mode = ((req->fs.info.mode & S_IRWXG) >> 3) | u_mode;
+  DWORD o_mode = ((req->fs.info.mode & S_IRWXO) >> 0) | g_mode;
+  build_explicit_access(&ea[0], psidOwner,    TRUSTEE_IS_USER,  u_mode);
+  build_explicit_access(&ea[2], psidGroup,    TRUSTEE_IS_GROUP, g_mode);
+  build_explicit_access(&ea[4], psidEveryone, TRUSTEE_IS_GROUP, o_mode);
+
+  /* Iterate over all old ACEs, looking for groups that we belong to */
+  ULONG ea_write_idx = 6;
+  for (ULONG ea_idx=0; ea_idx<numOldEAs; ++ea_idx) {
```

Comment:
```suggestion
  ULONG ea_idx;
  for (ea_idx = 0; ea_idx < numOldEAs; ea_idx++) {
```
(not permitted in C89)